content/blog/2020-11-29-tracemonkey/index.md

```diff
@@ -0,0 +1,402 @@
+boost of more than 5x, and two of those were 25x. Clearly TraceMonkey
+is a big improvement over SpiderMonkey already, but this was not at
+all unexpected.
+
+The true test of the tracing approach with type specialization is how
+it compares to a method-based JIT like V8 and an efficient interpreter
+like SFX. TraceMonkey is the fastest engine on nine of the benchmarks,
+and performs worst on thirteen benchmarks. Yet most of the
+performances are comparable to those of V8 and SFX; thus the authors'
+work is successful because it is a valid alternative to the existing
+efficient approaches.
+
+Critically, the poor performances of TraceMonkey are due primarily to
+limitations of its implementation that the authors planned to
+fix. These limitations include the issue with trace stitching
+described above, as well as the inability to trace across recursive
+function calls, into regular expression subtitutions with lambda
+functions, or through the usage of the `eval` function.
+
+The following graph shows the fraction of time that the benchmarks
+spent in each state from the high-level state machine:
+
+![breakdown of time spent in each state](results_breakdown.png)
+
+As can be seen, the benchmarks that performed the best spent almost
+all of their time running compiled traces, while those that performed
+the worse either were unable to compile any traces at all (due to
+languages features where tracing was not yet supported) or got bogged
+down tracing/compiling (possibly due to the issue with trace
+stitching). Thus the authors suspect that the performance will be
+superior and perhaps exceed that of V8 or SFX once these issues are
+ironed out. This paper is largely a proof of concept for tracing JITs
+with type specialization, and it is not necessarily expected to
+produce an implementation without quirks.
+
+## Discussion
+
+TraceMonkey was
+[merged](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Releases/1.8)
+into SpiderMonkey in version 1.8.1 and
+[released](https://wiki.mozilla.org/JavaScript:TraceMonkey) as part of
+Firefox 3.5. Clearly the work was deemed successful and, eventually,
+production-ready.
+
+That being said, TraceMonkey was
+[replaced](https://wiki.mozilla.org/JaegerMonkey) by a different
+implementation, JaegerMonkey, in 2010. JaegerMonkey is a method-based
+JIT with type specialization, rather than a tracing JIT. The current
+implementation is [IonMonkey](https://wiki.mozilla.org/IonMonkey),
+which is also a method-based JIT with type specialization. It seems
+that the tracing never caught on, but the type specialization did.
+
+The two main contributions of this paper are:
+
+1. how to create a tracing JIT for a dynamic language like JavaScript
+
+2. how to integrate type specialization in a JIT for a dynamic
+   language like JavaScript
+
+While both are very valuable, it seems that the type specialization is
+the more durable contribution.
```

Comment:
Indeed. For what it's worth, I still see it as an open question why TraceMonkey couldn't meet the same performance as JaegerMonkey. I have heard (just rumors) that the compiler was just too complicated—it wasn't a fundamental shortcoming in the tracing approach. But maybe there is some science to be done here to ask that question.